this addresses issue #22 by using the key when no translation is found

mvolkmann opened the pull request.

mvolkmann commented:
I signed it!

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -167,6 +179,20 @@
+
+    for (const instance of appLocalizeBehaviorInstances) {
+      // Update the language of all instances to match this one.
+      instance.language = this.language;
```
Hmm, this seems unrelated to the problem you're fixing, and slightly way bigger -- this basically makes it impossible to keep elements that use this behaviour use different languages at the same time. I don't think setting a language is that complicated or verbose, so I'd lean towards not doing this (certainly not doing it in this PR)

mvolkmann replied:
Sorry, I didn't mean to push that. I intended for that to be a separate pull request for a separate issue. You can ignore this commit.

I would however like the option to have all instances use the same language. I think that will be common case for most apps.

mvolkmann replied:
I will push a new version soon that removes the "language change" code.

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -194,6 +220,16 @@
   },
 
+  ready() {
+    console.log('app-localize-behavior.html ready: this.useKeyIfMissing =', this.useKeyIfMissing);
```
No `console.log`s, please

mvolkmann replied:
Removed in next version.

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -194,6 +220,16 @@
+    console.log('app-localize-behavior.html ready: this.useKeyIfMissing =', this.useKeyIfMissing);
+    // If the instance has the languageChanged method, call it.
+    // This allows it to set its initial state based on the language.
+    if (this.languageChanged) this.languageChanged();
```
If statements on a separate line, with `{}` please

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -210,20 +246,30 @@
       // do extra work if we're just reusing strings across an application.
-      var messageKey = key + resources[language][key];
-      var msg = proto.__localizationCache.messages[messageKey];
+      var value = resources[language][key];
```
can we name this something better than value? perhaps `translatedValue`?

mvolkmann replied:
trnaslatedValue and translatedMessage sound good. The name "msg" was already being used, but I'll change it to "translatedMessage".

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -210,20 +246,30 @@
+      if (value && !msg) {
+        try {
+          msg = new IntlMessageFormat(value, language, formats);
+          if (msg) proto.__localizationCache.messages[messageKey] = msg;
```
`If` statements on a new line, with `{}` please

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -210,20 +246,30 @@
+          msg = new IntlMessageFormat(value, language, formats);
+          if (msg) proto.__localizationCache.messages[messageKey] = msg;
+        } catch (e) {
+          if (!this.useKeyIfMissing) throw e;
```
Does this throw? When does this throw? I think adding a comment about this would be nice.

Also, why not assign `msg` to `key` in this block, rather than doing another check in L270?

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -115,7 +116,9 @@
     language: {
-      type: String
+      type: String,
+      value: Polymer.defaultLanguage,
```
Oof, I don't know about this. We set default values fairly differently -- the element that implements this behaviour can set a `value` for this attribute in its properties block. I don't think this is a good pattern

notwaldorf commented on demo/index.html:
```diff
@@ -17,13 +17,20 @@
+    <!-- Must import polymer before running next script. -->
+    <link rel="import" href="../../polymer/polymer.html">
+    <!-- Set this before importing components that use app-localize-behavior. -->
+    <script>Polymer.defaultLanguage = 'fr'</script>
```
As mentioned, I'm not super keen on this, as this is not how we set default values on elements

notwaldorf commented on demo/x-local-translate.html:
```diff
@@ -45,12 +45,6 @@
       properties: {
         /* Overriden from AppLocalizeBehavior */
-        language: {
-          value: 'en',
```
This was already setting a default value of `en` on this element -- I'm not sure I understand why this didn't work for you. You could also do 
```
language: {
  value: Polymer.defaultLanguage
}
```
here in your app if you wanted (but note, not in the behaviour itself like you've done it)

notwaldorf commented:
I think there's a bunch of problems with this PR right now -- I don't think I understand the reasoning behind changing how the default values are specified, nor the bit about assuming everyone has a `languageChanged` method on the element implementing this behaviour -- those both look like changes you want to use in _your_ element implementing the behaviour, not the behaviour itself.

The change in ` __computeLocalize` looks ok, but needs a bit more cleanup.

Let me know when this is ready for another look!

mvolkmann commented:
Okay, I think this is ready for another look now. I removed all the changes related to support global language setting and simplified my changes for useKeyIfMissing.

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -115,7 +115,7 @@
      */
     language: {
-      type: String
+      type: String,
```
no trailing comma, please

mvolkmann replied:
fixed in next push

notwaldorf commented on demo/x-local-translate.html:
```diff
@@ -50,7 +50,6 @@
           type: String
         },
 
-        /* Overriden from AppLocalizeBehavior */
```
I think this comment is still good :)

mvolkmann replied:
The comment is added back in my next commit.

notwaldorf commented on app-localize-behavior.html:
```diff
@@ -157,6 +157,15 @@
 
     /**
+     * When there is no translation for a given key,
+     * the key will be used when this is true.
```
How about rephrasing to "If true, will use the provided key when the translation does not exist for that key"

mvolkmann replied:
The comment is changed to your suggestion in my next push.

notwaldorf commented:
This looks good -- just a couple of nits left. Also, would you mind squashing your commits, please?

mvolkmann commented:
I made the changes you suggested and squashed the commits. It was my first time squashing commits. Hope I did it correctly.

mvolkmann commented:
Ignore this pull request. It has been superceded by https://github.com/PolymerElements/app-localize-behavior/pull/43
